contracts/ibc-reflect/src/contract.rs

```diff
@@ -0,0 +1,198 @@
+use cosmwasm_std::{entry_point, to_binary, DepsMut, Env, HandleResponse, HumanAddr, IbcAcknowledgement, IbcBasicResponse, IbcChannel, IbcOrder, IbcPacket, IbcReceiveResponse, InitResponse, MessageInfo, StdError, StdResult, WasmMsg, from_binary, CosmosMsg};
+
+use crate::msg::{HandleMsg, InitMsg, ReflectInitMsg, PacketMsg, AcknowledgementMsg, ReflectHandleMsg};
+use crate::state::{accounts, config, Config};
+
+const IBC_VERSION: &str = "ibc-reflect";
+
+#[entry_point]
+pub fn init(deps: DepsMut, _env: Env, _info: MessageInfo, msg: InitMsg) -> StdResult<InitResponse> {
+    // we store the reflect_id for creating accounts later
+    let cfg = Config {
+        reflect_code_id: msg.reflect_code_id,
+    };
+    config(deps.storage).save(&cfg)?;
+
+    Ok(InitResponse::default())
+}
+
+#[entry_point]
+pub fn handle(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    msg: HandleMsg,
+) -> StdResult<HandleResponse> {
+    match msg {
+        HandleMsg::InitCallback { id, contract_addr } => {
+            handle_init_callback(deps, info, id, contract_addr)
+        }
+    }
+}
+
+pub fn handle_init_callback(
+    deps: DepsMut,
+    info: MessageInfo,
+    id: String,
+    contract_addr: HumanAddr,
+) -> StdResult<HandleResponse> {
+    // sanity check - the caller is registering itself
+    if info.sender != contract_addr {
+        return Err(StdError::generic_err("Must register self on callback"));
+    }
+
+    // store id -> contract_addr if it is empty
+    // id comes from: `let chan_id = msg.endpoint.channel_id;` in `ibc_channel_connect`
+    accounts(deps.storage).update(id.as_bytes(), |val| -> StdResult<_> {
+        match val {
+            Some(_) => Err(StdError::generic_err(
+                "Cannot register over an existing channel",
+            )),
+            None => Ok(contract_addr),
+        }
+    })?;
+
+    Ok(HandleResponse::default())
+}
+
+#[entry_point]
+/// enforces ordering and versioing constraints
+pub fn ibc_channel_open(_deps: DepsMut, _env: Env, msg: IbcChannel) -> StdResult<()> {
+    if msg.order != IbcOrder::Ordered {
+        return Err(StdError::generic_err("Only supports ordered channels"));
+    }
+    if msg.version.as_str() != IBC_VERSION {
+        return Err(StdError::generic_err(format!(
+            "Must set version to `{}`",
+            IBC_VERSION
+        )));
+    }
+
+    Ok(())
+}
+
+#[entry_point]
+/// once it's established, we create the reflect contract
+pub fn ibc_channel_connect(
+    deps: DepsMut,
+    _env: Env,
+    msg: IbcChannel,
+) -> StdResult<IbcBasicResponse> {
+    let cfg = config(deps.storage).load()?;
+    let chan_id = msg.endpoint.channel_id;
+    let label = format!("ibc-reflect-{}", &chan_id);
+
+    let payload = to_binary(&ReflectInitMsg {
+        callback_id: Some(chan_id),
+    })?;
+    let msg = WasmMsg::Instantiate {
+        code_id: cfg.reflect_code_id,
+        msg: payload,
+        send: vec![],
+        label: Some(label),
```

Comment:
This is currently not supported in wasmd, right? https://github.com/CosmWasm/wasmd/blob/v0.14.0/x/wasm/internal/keeper/handler_plugin.go#L187-L188

Reply:
Good point. I just added https://github.com/CosmWasm/wasmd/issues/376